File: sach/de3327.py
//DE3327.PY - Tran Huu Nam - hsgtin.vn - 2024
import sys
from collections import defaultdict
n = 0
A = [0] * 1000002
P = defaultdict(list)
s = ""
def tim(x):
    global A, n
    A[0] = 0
    A[n + 1] = 0
    for i in range(1, n + 1):
        A[i] = A[i - 1]
        if s[i - 1] == x:
            A[i] += 1
    H = -n - 1
    cmin = n + 1
    Ci = j = t = 0
    for y in range(ord('a'), ord('z') + 1):
        if chr(y) == x or len(P[chr(y)]) == 1:
            continue
        cmin = n + 1
        for i in range(len(P[chr(y)])):
            j = P[chr(y)][i]
            Ci = A[j - 1] - i
            if cmin <= n:
                H = max(H, Ci - cmin)
            t = P[chr(y)][i - 1]
            cmin = min(cmin, A[t] - i)
    return H
def main():
    global n, s, P
    input = sys.stdin.read
    data = input().split()
    n = int(data[0])
    s = data[1]
    for i in range(n):
        P[s[i]].append(i + 1)
    for x in range(ord('a'), ord('z') + 1):
        P[chr(x)].append(n + 1)
    dmax = 0
    for x in range(ord('a'), ord('z') + 1):
        dmax = max(dmax, tim(chr(x)))
    print(dmax)
if __name__ == "__main__":
    main()

File: sach/de3327.cpp
//DE3327.CPP - Tran Huu Nam - hsgtin.vn - 2024
#include <bits/stdc++.h>
using namespace std;
int n, A[1000001];
vector <int> P[256]; string s;
int tim(char x) {
    A[0]=0; A[n+1]=0;
    for (int i=1; i<=n; i++) {
        A[i]=A[i-1];
        if (s[i-1]==x) A[i]++;
    }
    int H=-n-1, cmin, Ci, j, t;
    for (char y='a'; y<='z'; y++) {
        if (y==x || P[y].size()==1) continue;
        cmin=n+1;
        for (int i=0; i<P[y].size(); i++) {
            j=P[y][i]; Ci = A[j-1] - i;
            if (cmin<=n) H=max(H, Ci - cmin);
            t=P[y][i-1];
            cmin=min(cmin, A[t]-i);
        }
    }
    return H;
}
int main() {
    cin>>n>>s;
    for (int i=0; i<n; i++)
        P[s[i]].push_back(i+1);
    for (char x='a'; x<='z'; x++)
      P[x].push_back(n+1);
    int dmax=0;
    for (char x='a'; x<='z'; x++)
        dmax=max(dmax, tim(x));
    cout<<dmax;
    return 0;
}

